### PR TITLE
Allow OAuth application client secret reset by PATCH [PLAT-1163]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose
A cleaner way to reset the client secret for an OAuth application via the API would be to PATCH the developer-app resource with the client_secret property set to null. This would be more intuitive that POSTing to a special endpoint and would make the front end simpler.

**Note:** This uses version 2.13 as the "new version," but there are a bunch of new API versions in currently open PRs and so the version can be adjusted when this PR gets close to being merged.

## Changes
- Deprecate old ApplicationReset view
- allow patching client_secret as null in ApiOAuth2ApplicationDetailSerializer
- Tests for deprecated view and new route

## QA Notes
This is purely API related, so assuming it'll be tested by spot checking and the front-end team when they use it!

## Documentation
It doesn't look like any of the current application API endpoints are currently in the documentation, so no updates needed until we decide to document the applications endpoints.

## Side Effects
None anticipated

## Ticket

https://openscience.atlassian.net/browse/PLAT-1163

### CODE REVIEW

**Comment by Johnetordoff** [comment]
Yeah this honestly looks perfect, I wounldn't change a thing

**Comment by Johnetordoff** [comment]
Yeah this honestly looks perfect, I wounldn't change a thing

File: api/applications/serializers.py [hunk 1]
```diff
@@ -37,12 +39,13 @@ def get_absolute_url(self, obj):
...
-                'version': self.context['request'].parser_context['kwargs']['version'],
-            },
-        )
+        if StrictVersion(self.context['request'].version) < StrictVersion('2.13'):
+            return absolute_reverse(
+                'applications:application-reset', kwargs={
+                    'client_id': obj.client_id,
+                    'version': self.context['request'].parser_context['kwargs']['version'],
+                },
+            )
```
Comment (pattisdr) [inline api/applications/serializers.py]: 👍 reset_url only returned in early versions (here post-2.13, but will need to be updated w/ the latest version).  Later versions, reset_url will not be serialized with the application at all.

File: api/applications/serializers.py [hunk 2]
```diff
@@ -99,8 +102,26 @@ def create(self, validated_data):
...
+        help_text='The client secret for this application (automatically generated)',
+        allow_null=True,
+        required=False,
+    )
+
     def update(self, instance, validated_data):
         assert isinstance(instance, ApiOAuth2Application), 'instance must be an ApiOAuth2Application'
+        client_secret = validated_data.pop('client_secret', None)
+        if client_secret == '':
+            instance.reset_secret(save=True)
```
Comment (pattisdr) [inline api/applications/serializers.py]: To reset the secret, request body will now look like this:
```json
{
	"data": {
		"id": "application_idfkdsljfjasd",
		"type": "applications",
		"attributes": {
		   "client_secret": null
		}
	}
}
```
with allow_null is true, `null` will be converted to an empty string after data is validated.  allow_blank is not true, so an empty string can't be passed in into the request

File: api/applications/views.py [hunk 3]
```diff
@@ -104,12 +104,15 @@ def perform_update(self, serializer):
...
-class ApplicationReset(JSONAPIBaseView, generics.CreateAPIView, ApplicationMixin):
+class ApplicationReset(DeprecatedView, generics.CreateAPIView, ApplicationMixin):
     """
     Resets client secret of a specific API application (eg OAuth2) that the user has registered
 
     Should not perform update or return information if the application belongs to a different user
     """
+
+    max_version = '2.12'
+
```
Comment (pattisdr) [inline api/applications/views.py]: will increase the max_version to 2.14.

File: api_tests/applications/views/test_application_reset.py [hunk 4]
```diff
@@ -86,18 +107,78 @@ def test_other_user_cannot_reset(
...
         assert old_secret == user_app.client_secret
 
-    @mock.patch('osf.models.ApiOAuth2Application.reset_secret')
-    def test_unauth_user_cannot_reset(
-            self, mock_method, app, user_app, user_reset_url, correct
-    ):
-        mock_method.return_value(True)
-        old_secret = user_app.client_secret
+        # unauthorized user reset fails
         res = app.post_json_api(
```
Comment (pattisdr) [inline api_tests/applications/views/test_application_reset.py]: should be patch_json_api

File: api_tests/applications/views/test_application_reset.py [hunk 5]
```diff
@@ -59,25 +76,29 @@ def test_reset_does_not_save_without_save_param(
...
+        assert 'This route is deprecated' in res.json['meta']['warnings'][0]
         mock_method.assert_called_with(user_app.client_id, old_secret)
         user_app.reload()
         assert old_secret != user_app.client_secret
 
     @mock.patch('osf.models.ApiOAuth2Application.reset_secret')
-    def test_other_user_cannot_reset(
-            self, mock_method, app, user_app, user_reset_url, correct
-    ):
+    def test_reset_fails(
```
Comment (pattisdr) [inline api_tests/applications/views/test_application_reset.py]: would retain permissions tests for deprecated view as it still exists.

**Comment by pattisdr** [comment]
replaced by https://github.com/CenterForOpenScience/osf.io/pull/8993

**Pull request closed without merging**